gnomad/utils/file_utils.py

```diff
@@ -280,3 +280,44 @@ def read_list_data(input_file_path: str) -> List[str]:
         output.append(line.strip())
     f.close()
     return output
+
+
+def repartition_for_join(
+    ht_path: str,
+    new_partition_percent: Optional[float] = None,
+    n_partitions: Optional[int] = None,
+) -> hl.Table:
+    """
+    Repartition a Table prior to joining with another Table(s).
+
+    This repartitioning makes the join(s) much more efficient.
+
+    :param ht_path: Path to Table to repartition.
+    :param new_partition_percent: Percent of initial dataset partitions to use.
+        Value should be greater than 1.
+        Default is None.
+        Should not be specified if `n_partitions` is not None.
+    :param n_partitions: Number of partitions desired after repartition.
+        Default is None.
+        Should not be specified if `new_partition_percent` is not None.
+    :return: hl.Table
+    """
+    if not new_partition_percent and not n_partitions:
+        raise DataException(
+            "Must specify either new_partition_percent or n_partitions!"
+        )
+
+    ht = hl.read_table(ht_path)
+    if new_partition_percent:
+        if new_partition_percent < 1:
+            logger.warning(
+                "new_partition_percent value is less than 1! The new HT will have fewer"
+                " partitions than the original HT!"
+            )
+        partition_intervals = ht._calculate_new_partitions(
+            ht.n_partitions() * new_partition_percent
+        )
+    if n_partitions:
+        partition_intervals = ht._calculate_new_partitions(n_partitions)
+
+    return hl.read_table(ht_path, _intervals=partition_intervals)
```

Comment:
So I think the script should return the partition_intervals(an array of intervals, one entry per partition) and the  original table re-read or just the partition_intervals. The performance gain we see in a join is because all the tables involved are partitioned using the same intervals. Right now, I don't know of a way to access the partition intervals cleanly from the returned table.

Reply:
just to make sure I'm following, there are two thoughts here:

- the return type should be both the Table and the newly calculated intervals (hl.Table, hl.expr.ArrayExpression) or just the intervals?
- `n_partitions` isn't helpful for the join here?

I'm happy to change, though when I was doing quick joins for MPC, I actually just set the new partitions using `n_partitions` (I think I set them all to 15k with `_calculate_new_partitions`), and that sped up the join more than me simply repartitioning each HT on read to 15k partitions. Not sure if that wasn't best practices

Reply:
> the return type should be both the Table and the newly calculated intervals (hl.Table, hl.expr.ArrayExpression) or just the intervals?

Yeah so my understanding to what matters for the join efficiency is the intervals covered within each partition so returning the intervals would be most useful for joining because you then just use that in 
```
ht=hl.read_table(<path>, _intervals=partition_intervals)
```
The _intervals size is equal to n_partitions in this case so youre getting that for free... just getting extra info for where the table should be diced up.

>n_partitions isn't helpful for the join here?

It is helpful but not as helpful as `_intervals`. When you pass 15k to `_calculate_new_partitions` for table A you're saying I want 15k partitions on table A but that doesnt mean you'll get the same intervals in each partition for when you call it on table B. Since using `hl.read_table(<path>, _n_partitions=n_partitions)` calls `_calculcate_new_partitions` in hail's `read_table`

```
    if _intervals is not None and _n_partitions is not None:
        raise ValueError("'read_table' does not support both _intervals and _n_partitions")
    tr = ir.TableNativeReader(path, _intervals, _filter_intervals)
    ht = Table(ir.TableRead(tr, False, drop_row_uids=not _create_row_uids, _assert_type=_assert_type))

    if _n_partitions:
        intervals = ht._calculate_new_partitions(_n_partitions)
        return read_table(path, _intervals=intervals, _assert_type=_assert_type, _load_refs=_load_refs, _create_row_uids=_create_row_uids)
    return ht
```
 I'm guessing the tables are similar sizes for MPC? So `n_parititons` is probably making the intervals  close and speeding things up. The _intervals just gives us  the added benefit of partitions matching their start and end locus when the tables are mismatched. Sorry if I went extra basic here, maybe lets have a meeting if I'm not being clear?

Reply:
No this is great, thanks! Will update.

One quick note:
> and I'm guessing the tables are similar sizes for MPC

They actually aren't, but this might be an instance where spark partitioning is a black box, and I'm not sure why/how using only `_n_partitions` helped speed up the joins so much

Reply:
Huh, my guess is it is a parallelization boost in spark? but yeah -- black box